consensus/istanbul/backend/announce.go

```diff
@@ -700,3 +719,175 @@ func (sb *Backend) checkPeersAnnounceVersions() {
+// validator to its proxy.
+func (sb *Backend) handleVersionedEnodeMsg(peer consensus.Peer, payload []byte) error {
+	logger := sb.logger.New("func", "handleVersionedEnodeMsg")
+
+	// TODO remove proxy checks in this function once versioned enode messages are
+	// not limited proxied validator -> proxy communication
+	// Issue tracked here: https://github.com/celo-org/celo-blockchain/issues/884
+	if !sb.config.Proxy {
+		return errNotProxy
+	}
+
+	var msg istanbul.Message
+	// Decode payload into msg
+	err := msg.FromPayload(payload, istanbul.GetSignatureAddress)
+	if err != nil {
+		logger.Error("Error in decoding received Istanbul Versioned Enode message", "err", err, "payload", hex.EncodeToString(payload))
+		return err
+	}
+	logger.Trace("Handling an Istanbul Versioned Enode message", "from", msg.Address)
+
+	if msg.Address != sb.config.ProxiedValidatorAddress {
+		logger.Error("Error in the origin of a received Istanbul Versioned Enode message", "msg address", msg.Address.String(), "proxied validator address", sb.config.ProxiedValidatorAddress.String())
+		return fmt.Errorf("Origin address is not the proxied validator (%s != %s)", msg.Address.String(), sb.config.ProxiedValidatorAddress.String())
+	}
+
+	var versionedEnode versionedEnode
+	if err := rlp.DecodeBytes(msg.Msg, &versionedEnode); err != nil {
+		logger.Warn("Error in decoding received Istanbul Versioned Enode message content", "err", err, "IstanbulMsg", msg.String())
+		return err
+	}
+
+	parsedNode, err := enode.ParseV4(versionedEnode.Node)
+	if err != nil {
+		logger.Warn("Malformed v4 node in received Istanbul Versioned Enode message", "versionedEnode", versionedEnode, "err", err)
+	}
+
+	// There may be a difference in the URLv4 string because of `discport`,
+	// so instead compare the ID
+	selfNode := sb.p2pserver.Self()
+	if parsedNode.ID() != selfNode.ID() {
+		err := fmt.Errorf("Unexpected node (%s != %s)", versionedEnode.Node, sb.p2pserver.Self().URLv4())
+		logger.Warn("Received Istanbul Versioned Enode message with an incorrect node", "err", err)
+		return err
+	}
+
+	logger.Trace("Received Istanbul Versioned Enode message", "versionedEnode", versionedEnode)
+
+	sb.selfVersionedEnodeMsgMu.Lock()
+	sb.selfVersionedEnodeMsg = &msg
+	sb.selfVersionedEnodeMsgMu.Unlock()
+
+	return nil
+}
+
+func (sb *Backend) sendVersionedEnodeMsg(peer consensus.Peer, version uint) error {
+	logger := sb.logger.New("func", "sendVersionedEnodeMsg")
+
+	versionedEnodeMsg, err := sb.retrieveSelfVersionedEnodeMsg(version)
+	if err != nil {
+		logger.Warn("Error getting self versioned enode message", "err", err)
+		return err
+	}
+	if versionedEnodeMsg == nil {
+		return nil
+	}
+
+	payload, err := versionedEnodeMsg.Payload()
+	if err != nil {
+		logger.Error("Error getting payload of versioned enode message", "err", err)
+		return err
+	}
+	return peer.Send(istanbulVersionedEnodeMsg, payload)
+}
+
+func getCurrentAnnounceVersion() uint {
```

Comment:
Should the announce version always be updated here?

Reply:
I just refactored a little to remove the notion of a "fallback version", so now every call to `getCurrentAnnounceVersion` is when the announce version is being updated. I'm a little confused what you mean here